src/caffe/test/test_reduction_layer.cpp

```diff
@@ -0,0 +1,209 @@
+#include <algorithm>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+#include "caffe/blob.hpp"
+#include "caffe/common.hpp"
+#include "caffe/filler.hpp"
+#include "caffe/vision_layers.hpp"
+
+#include "caffe/test/test_caffe_main.hpp"
+#include "caffe/test/test_gradient_check_util.hpp"
+
+namespace caffe {
+
+template <typename TypeParam>
+class ReductionLayerTest : public MultiDeviceTest<TypeParam> {
+  typedef typename TypeParam::Dtype Dtype;
+
+ protected:
+  ReductionLayerTest()
+      : blob_bottom_(new Blob<Dtype>(2, 3, 4, 5)),
+        blob_top_(new Blob<Dtype>()) {
+    // fill the values
+    Caffe::set_random_seed(1701);
+    FillerParameter filler_param;
+    UniformFiller<Dtype> filler(filler_param);
+    filler.Fill(this->blob_bottom_);
+    blob_bottom_vec_.push_back(blob_bottom_);
+    blob_top_vec_.push_back(blob_top_);
+  }
+  virtual ~ReductionLayerTest() {
+    delete blob_bottom_;
+    delete blob_top_;
+  }
+
+  void TestForward(ReductionParameter_ReductionOp op, float coeff = 1) {
+    LayerParameter layer_param;
+    ReductionParameter* reduction_param = layer_param.mutable_reduction_param();
+    reduction_param->set_operation(op);
+    if (coeff != 1.0) {
+      reduction_param->set_coeff(2.3);
```

Comment:
`reduction_param->set_coeff(coeff);` might be more robust for cases when `coeff` != 2.3.